### PR TITLE
feature(hx): add cross-domain scenario.

### PR DESCRIPTION
**Pull request by XinHao-96**
A pipeline for federated learning in cross-domain scenarios. In this setup, client data originates from different domains (it should be noted that in the cross-domain pipeline, the hyperparameter --client_num refers to the number of clients into which data from the same domain has been divided). Final training performance is evaluated locally on each client (it is important to note that clients within the same domain share the same test data).

Dataset: A new function, `get_cross_domain_dataset`, has been created to load datasets based on domain information. Additionally, a new cross-domain dataset, DomainNet, has been included.

Data Sampling: A conditional check for cross-domain datasets has been added, ensuring that clients from the same domain share the same test set.

Group: A new class, `CrossDomainServerGroup`, has been created, inheriting from the `ParameterServerGroup` parent class. The primary modification involves indexing clients based on domain information and the number of clients per domain. To support this, a new `cross-domain-avg` method was also implemented to enable proper aggregation for clients with domain-specific information.

Server: A new class, `CrossDomainServer`, was created, inheriting from `ServerTemplate`. The primary modification involves adding indexing based on domain information (the server testing method from `generic_model_pipeline` can be used here).

Client: To enable precise localization of cross-domain clients, a combined indexing method using both domain name and client ID has been implemented. A new function, `get_cross_domain_client`, was created to obtain these two parameters. Additionally, a `CrossDomainClient` class, designed for cross-domain scenarios, was introduced, inheriting from `ClientTemplate`.
                                                                                                                                                                                          --Xin Hao

### CODE REVIEW

File: fling/component/client/build_client.py [hunk 1]
```diff
@@ -15,3 +15,15 @@ def get_client(
         test_dataset=test_dataset,
         **kwargs
     )
+def get_cross_domain_client(
```
Comment (kxzxvbk) [inline fling/component/client/build_client.py]: Maybe there is no need to add this function. You can use ``**kwargs`` to pass the argument domain.

File: fling/component/client/cross_domain_client.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+from torch.utils.data import DataLoader, Dataset
+import torch.nn as nn
+
+from fling.utils import get_optimizer, VariableMonitor, get_weights
+from fling.utils.registry_utils import CLIENT_REGISTRY
+from .client_template import ClientTemplate
+
+
+@CLIENT_REGISTRY.register('cross_domain_client')
+class CrossDomainClient(ClientTemplate):
```
Comment (kxzxvbk) [inline fling/component/client/cross_domain_client.py]: Maybe this class can inherit BaseClient. I think there is no difference in function train, test, finetune, preprocess_data. And the __init__ function can be largely simplified by calling super.__init__()

File: fling/component/group/cross_domain_group.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        self.sync()
+
+        # Logging model information.
+        self.logger.logging(str(self.clients[self.domains[0]][0].model))
+        self.logger.logging('All clients initialized.')
+        self.logger.logging(
+            'Parameter number in each model: {:.2f}M'.format(get_params_number(self.clients[self.domains[0]][0].model) / 1e6)
+        )
+    
+    def append(self, domain: str, client: ClientTemplate) -> None:
```
Comment (kxzxvbk) [inline fling/component/group/cross_domain_group.py:65]: Remove this function?

File: fling/component/group/cross_domain_group.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        self.logger.add_scalar('time/time_per_round', time_per_round, train_round)
+
+        if aggr_parameter_args is not None:
+            for domain in self.domains:
+                for client in range(self.num_user):
+                    self.clients[domain][client].set_fed_keys(fed_keys_bak)
+
+        return trans_cost
+
+    def flush(self) -> None:
```
Comment (kxzxvbk) [inline fling/component/group/cross_domain_group.py]: Remove this function?

File: fling/component/server/cross_domain_server.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        
+        monitor.append(
+            {
+                f'{domain}_test_acc': torch.mean((y_pred == batch_y).float()).item(),
+                f'{domain}_test_loss': loss.item()
+            },
+            weight=batch_y.shape[0]
+        )
+    
+    def preprocess_data(self, data):
```
Comment (kxzxvbk) [inline fling/component/server/cross_domain_server.py]: Maybe remove this function.

File: flzoo/domainnet/domainnet_fedavg_resnet_config.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    ),
+    learn=dict(
+        device='cuda:0', local_eps=1, global_eps=200, batch_size=64, optimizer=dict(name='sgd', lr=0.01, momentum=0.5)
+    ),
+    model=dict(
+        name='resnet8',
+        input_channel=3,
+        class_number=10,
+    ),
+    client=dict(name='cross_domain_client', client_num=2),
```
Comment (kxzxvbk) [inline flzoo/domainnet/domainnet_fedavg_resnet_config.py]: Give more detailed comments about the meaning of `client_num` here, as it is not the same meaning as that in typical FL.

**Comment by kxzxvbk** [comment]
Reformat your code by:
```
source ./format.sh ./
flake8 ./
```

**Comment by kxzxvbk** [comment]
Add cross-domain information in the README file.

**Comment by XinHao-96** [comment]
Based on your suggestions, I have made the following modifications to the original code:

Changed `cross_domain_client` to inherit from `BaseClient` and removed redundant parts.
Removed redundant sections in `cross_domain_group`.
Changed `cross_domain_server` to inherit from `BaseServer` and eliminated duplicate code.
Reformatted the parts related to cross-domain modifications using `source ./format.sh ./`. During the actual run, I also made formatting adjustments to other parts, such as the `FedPart` code, but to avoid confusion, I have not reflected these changes in the pull request.

One part that still confuses me is that the Fling README file does not seem to contain a description of heterogeneity or the scenarios. Therefore, I am unsure where it would be most appropriate to place the introduction to cross-domain.

File: fling/dataset/build_dataset.py [hunk 7]
```diff
@@ -5,3 +5,7 @@
 
 def get_dataset(args: dict, train: bool, **kwargs) -> Dataset:
     return DATASET_REGISTRY.build(args.data.dataset, cfg=args, train=train, **kwargs)
+
+
+def get_cross_domain_dataset(args: dict, domain: str, train: bool, **kwargs) -> Dataset:
```
Comment (kxzxvbk) [inline fling/dataset/build_dataset.py]: Maybe this function can be directly removed, and simply use ``get_dataset``?

**Comment by XinHao-96** [comment]
removed the get_cross_domain_dataset function by using get_dataset function instead.
i'm still working on enlarging readme file.

**Comment by XinHao-96** [comment]
removed the get_cross_domain_dataset function by using get_dataset function instead.
i'm still working on enlarging readme file.

**Comment by XinHao-96** [comment]
I enlarged the cross-domain description part in readme.md and created two demos by using FedPer and FedBN algorithm with DomainNet dataset.

**Comment by XinHao-96** [comment]
I followed the instructions in the file to download DomainNet, and then I ran the test locally. It worked fine.
